In [1]:
# Importing Necessary Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Importing Dataset
df = pd.read_csv("salary.csv")
pd.set_option("display.max_columns", None) 
df.head(1)

C:\Users\sweet\AppData\Local\Temp\ipykernel_7156\3210461817.py:2: DtypeWarning: Columns (16,18,20) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("salary.csv")


AGENCY                                        AGENCY NAME  \
0     241  COMPTROLLER OF PUBLIC ACCOUNTS, JUDICIARY SECT...   

                        LAST NAME                      FIRST NAME MI  \
0  RUCKER                          MORTON                          V   

  CLASS CODE                                        CLASS TITLE  \
0   JD25      JUDGE, RETIRED                                ...   

         ETHNICITY           GENDER                                    STATUS  \
0  WHITE            MALE             URP - UNCLASSIFIED REGULAR PART-TIME       

  EMPLOY DATE  HRLY RATE  HRS PER WK  MONTHLY     ANNUAL  STATE NUMBER  \
0    02/18/88    75.9615        29.0  9545.82  114549.84        127717   

  duplicated  multiple_full_time_jobs combined_multiple_jobs  \
0       True                      NaN                    NaN   

   summed_annual_salary hide_from_search  
0             131407.08              NaN

In [3]:
# There is a warning while loading the dataset this is due to the dataset cotains mixed data types to overcome this
# Loading the dataset with low_memory set to False
df = pd.read_csv("salary.csv", low_memory=False)
df.head()

AGENCY                                        AGENCY NAME  \
0     241  COMPTROLLER OF PUBLIC ACCOUNTS, JUDICIARY SECT...   
1     212  OFFICE OF COURT ADMINISTRATION                ...   
2     241  COMPTROLLER OF PUBLIC ACCOUNTS, JUDICIARY SECT...   
3     212  OFFICE OF COURT ADMINISTRATION                ...   
4     696  TEXAS DEPARTMENT OF CRIMINAL JUSTICE          ...   

                        LAST NAME                      FIRST NAME MI  \
0  RUCKER                          MORTON                          V   
1  RUCKER                          MORTON                          V   
2  SPECIA JR                       JOHN                            J   
3  SPECIA JR                       JOHN                            J   
4  ONTIVEROS                       ESTHER                              

  CLASS CODE                                        CLASS TITLE  \
0   JD25      JUDGE, RETIRED                                ...   
1   3524      GENERAL COUNSEL IV                            ...   
2   JD25      JUDGE, RETIRED                                ...   
3   3524      GENERAL COUNSEL IV                            ...   
4   4504      CORREC  OFFICER IV                            ...   

         ETHNICITY           GENDER                                    STATUS  \
0  WHITE            MALE             URP - UNCLASSIFIED REGULAR PART-TIME       
1  WHITE            MALE             CTP - CLASSIFIED TEMPORARY PART-TIME       
2  WHITE            MALE             URP - UNCLASSIFIED REGULAR PART-TIME       
3  WHITE            MALE             CTP - CLASSIFIED TEMPORARY PART-TIME       
4  HISPANIC         FEMALE           CRF - CLASSIFIED REGULAR FULL-TIME         

  EMPLOY DATE  HRLY RATE  HRS PER WK  MONTHLY     ANNUAL  STATE NUMBER  \
0    02/18/88   75.96150        29.0  9545.82  114549.84        127717   
1    02/01/15   81.04454         4.0  1404.77   16857.24        127717   
2    02/01/20   75.96150        29.0  9545.82  114549.84         59115   
3    09/01/18   81.04453         4.0  1404.77   16857.24         59115   
4    06/29/20    0.00000        40.0  3284.27   39411.24        165030   

  duplicated  multiple_full_time_jobs combined_multiple_jobs  \
0       True                      NaN                    NaN   
1       True                      NaN                    NaN   
2       True                      NaN                    NaN   
3       True                      NaN                    NaN   
4       True                      1.0                    NaN   

   summed_annual_salary hide_from_search  
0             131407.08              NaN  
1                   NaN             True  
2             131407.08              NaN  
3                   NaN             True  
4                   NaN              NaN

# 1) Basic Checks and data filtering

In [9]:
df.shape

(149481, 21)

#### There are 21 features present in the dataset with each having 149481 entries.
####

In [12]:
# Checking for the duplicates 
df.duplicated().sum

<bound method Series.sum of 0         False
1         False
2         False
3         False
4         False
          ...  
149476    False
149477    False
149478    False
149479    False
149480    False
Length: 149481, dtype: bool>

#### No actual duplicates present in the data.
####

In [16]:
# Now let's check for the column 'duplicated'.
# Checking for unique entries in the 'duplicated' column.
print(df['duplicated'].unique())
# Checking the occurence of the unique values.
print(df['duplicated'].value_counts())

[True nan]
duplicated
True    143
Name: count, dtype: int64


In [18]:
# Checking for unique entries in 'multiple_full_time_jobs' column.
print(df['multiple_full_time_jobs'].unique())
# Checking the occurence of the unique values.
print(df['multiple_full_time_jobs'].value_counts())

[nan  1.]
multiple_full_time_jobs
1.0    14
Name: count, dtype: int64


In [20]:
# Checking for unique entries in 'combined_multiple_jobs' column.
print(df['combined_multiple_jobs'].unique())
# Checking for the occurence of unique values.
print(df['combined_multiple_jobs'].value_counts())

[nan True]
combined_multiple_jobs
True    97
Name: count, dtype: int64


In [22]:
# Checking for unique entries in 'summed_annual_salary' column.
print(df['summed_annual_salary'].nunique())
# Checking for the occurence of unique values.
print(df['summed_annual_salary'].value_counts())

15
summed_annual_salary
131407.08    2
171383.40    1
31950.00     1
33300.00     1
33966.00     1
39174.00     1
35570.52     1
56952.60     1
21839.76     1
121200.00    1
90000.00     1
150000.00    1
92400.00     1
18000.00     1
30312.84     1
Name: count, dtype: int64


In [24]:
# Checking for unique entries in 'hide_from_search' column.
print(df['hide_from_search'].nunique())
# Checking for the occurence of unique values.
print(df['hide_from_search'].value_counts())

1
hide_from_search
True    16
Name: count, dtype: int64


### Interpretation:
#### "The columns 'is_duplicated', 'multiple_full_time_jobs', 'combined_multiple_jobs', 'summed_annual_salary', and 'hide_from_search' do not contribute much to the dataset because most of their values are 'NaN'. Therefore, we will remove these columns
####

In [27]:
# Removing the unnecessary columns.
columns_to_remove = ['duplicated',  'multiple_full_time_jobs', 'combined_multiple_jobs', 
                     'summed_annual_salary', 'hide_from_search']
df = df.drop(columns=columns_to_remove)
df.head(2)

AGENCY                                        AGENCY NAME  \
0     241  COMPTROLLER OF PUBLIC ACCOUNTS, JUDICIARY SECT...   
1     212  OFFICE OF COURT ADMINISTRATION                ...   

                        LAST NAME                      FIRST NAME MI  \
0  RUCKER                          MORTON                          V   
1  RUCKER                          MORTON                          V   

  CLASS CODE                                        CLASS TITLE  \
0   JD25      JUDGE, RETIRED                                ...   
1   3524      GENERAL COUNSEL IV                            ...   

         ETHNICITY           GENDER                                    STATUS  \
0  WHITE            MALE             URP - UNCLASSIFIED REGULAR PART-TIME       
1  WHITE            MALE             CTP - CLASSIFIED TEMPORARY PART-TIME       

  EMPLOY DATE  HRLY RATE  HRS PER WK  MONTHLY     ANNUAL  STATE NUMBER  
0    02/18/88   75.96150        29.0  9545.82  114549.84        127717  
1    02/01/15   81.04454         4.0  1404.77   16857.24        127717

In [29]:
# Checking the data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149481 entries, 0 to 149480
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   AGENCY        149481 non-null  int64  
 1   AGENCY NAME   149481 non-null  object 
 2   LAST NAME     149481 non-null  object 
 3   FIRST NAME    149481 non-null  object 
 4   MI            149481 non-null  object 
 5   CLASS CODE    149481 non-null  object 
 6   CLASS TITLE   149481 non-null  object 
 7   ETHNICITY     149481 non-null  object 
 8   GENDER        149481 non-null  object 
 9   STATUS        149481 non-null  object 
 10  EMPLOY DATE   149481 non-null  object 
 11  HRLY RATE     149481 non-null  float64
 12  HRS PER WK    149481 non-null  float64
 13  MONTHLY       149481 non-null  float64
 14  ANNUAL        149481 non-null  float64
 15  STATE NUMBER  149481 non-null  int64  
dtypes: float64(4), int64(2), object(10)
memory usage: 18.2+ MB


#### There are total 6 columns with the 'numeric' data type and 10 columns with the 'object' data type.
####

In [32]:
# Renaming the features
new_column_names = {
    'AGENCY': 'agency',
    'AGENCY NAME': 'agency_name',
    'LAST NAME': 'last_name',
    'FIRST NAME': 'first_name',
    'MI': 'middle_initial',
    'CLASS TITLE': 'class_title',
    'CLASS CODE': 'class_code',
    'ETHNICITY': 'ethnicity',
    'GENDER': 'gender',
    'STATUS': 'status',
    'EMPLOY DATE': 'employ_date',
    'HRLY RATE': 'hourly_rate',
    'HRS PER WK': 'hours_per_week',
    'MONTHLY': 'monthly_income',
    'ANNUAL': 'annual_income',
    'summed_anual_salary': 'summed_annual_income',
    'STATE NUMBER': 'state_number'
}
df.rename(columns=new_column_names, inplace=True)
print(df.columns)

Index(['agency', 'agency_name', 'last_name', 'first_name', 'middle_initial',
       'class_code', 'class_title', 'ethnicity', 'gender', 'status',
       'employ_date', 'hourly_rate', 'hours_per_week', 'monthly_income',
       'annual_income', 'state_number'],
      dtype='object')


In [34]:
df.head(1)

agency                                        agency_name  \
0     241  COMPTROLLER OF PUBLIC ACCOUNTS, JUDICIARY SECT...   

                        last_name                      first_name  \
0  RUCKER                          MORTON                           

  middle_initial class_code  \
0              V   JD25       

                                         class_title        ethnicity  \
0  JUDGE, RETIRED                                ...  WHITE             

            gender                                    status employ_date  \
0  MALE             URP - UNCLASSIFIED REGULAR PART-TIME        02/18/88   

   hourly_rate  hours_per_week  monthly_income  annual_income  state_number  
0      75.9615            29.0         9545.82      114549.84        127717

# 2) EDA

## i) Descriptive Statistics

In [38]:
df.describe()

agency    hourly_rate  hours_per_week  monthly_income  \
count  149481.000000  149481.000000   149481.000000   149481.000000   
mean      545.736288       0.346488       39.714918     4226.184248   
std       139.057691       4.211485        2.323855     2093.111691   
min       101.000000       0.000000        1.000000       50.000000   
25%       529.000000       0.000000       40.000000     3019.840000   
50%       530.000000       0.000000       40.000000     3720.170000   
75%       696.000000       0.000000       40.000000     4865.810000   
max       930.000000     117.788480       70.000000    46125.000000   

       annual_income  state_number  
count  149481.000000  1.494810e+05  
mean    50714.210973  5.651930e+05  
std     25117.340292  4.786212e+05  
min       600.000000  1.000000e+00  
25%     36238.080000  1.133340e+05  
50%     44642.040000  3.463730e+05  
75%     58389.720000  1.100500e+06  
max    553500.000000  1.271604e+06